src/gen/typescript.rs

```diff
@@ -0,0 +1,546 @@
+use heck::*;
+use oasis_rpc::Interface;
+use proc_macro2::{Ident, Literal, TokenStream};
+use quote::{format_ident, quote};
+
+macro_rules! format_ts_ident {
+    (@import, $name:expr) => {
+        format_ts_ident!(@raw, $name.to_mixed_case())
+    };
+    (@class, $name:expr) => {
+        format_ts_ident!(@raw, $name.to_camel_case())
+    };
+    (@var, $name:expr) => {
+        format_ts_ident!(@raw, $name.to_mixed_case())
+    };
+    (@const, $name:expr) => {
+        format_ts_ident!(@raw, $name.to_shouty_snake_case());
+    };
+    (@raw, $name:expr) => {
+        format_ident!("{}", $name)
+    };
+}
+
+pub fn generate(iface: &Interface, bytecode_url: &url::Url) -> TokenStream {
+    let service_ident = format_ts_ident!(@class, iface.name);
+    let bytecode_url_str = bytecode_url.as_str();
+
+    let imports = iface.imports.iter().map(|imp| {
+        let import_ident = format_ts_ident!(@class, imp.name);
+        let module_name = module_name(&imp.name);
+        quote!(import  #import_ident from #module_name)
+    });
+
+    let type_defs = generate_type_defs(&iface.type_defs);
+
+    let deploy_function = generate_deploy_function(&service_ident, &iface.constructor);
+    let rpc_functions = generate_rpc_functions(&service_ident, &iface.functions);
+
+    quote! {
+        import { Buffer } from "buffer";
+        import * as oasis from "oasis-std";
+
+        #(#imports)*
+
+        #(#type_defs)*
+
+        export class #service_ident {
+            public static BYTECODE_URL = #bytecode_url_str;
+
+            private constructor(readonly address: oasis.Address, private gateway: oasis.Gateway) {}
+
```

Comment:
do we need a create method as well? Or is there a custom create method for each service?

Reply:
yes, that's `deploy`, which is `#deploy_function`